File: go/cache/pubsub.go
package cache

import "sync"

type Subscriber func(payload any)

type PubSub struct {
	mu   sync.RWMutex
	subs map[string][]Subscriber
}

func NewPubSub() *PubSub {
	return &PubSub{subs: map[string][]Subscriber{}}
}

func (p *PubSub) Subscribe(channel string, handler Subscriber) func() {
	p.mu.Lock()
	p.subs[channel] = append(p.subs[channel], handler)
	p.mu.Unlock()
	return func() {
		p.mu.Lock()
		defer p.mu.Unlock()
		handlers := p.subs[channel]
		for i, h := range handlers {
			if h == handler {
				p.subs[channel] = append(handlers[:i], handlers[i+1:]...)
				break
			}
		}
	}
}

func (p *PubSub) Publish(channel string, payload any) int {
	p.mu.RLock()
	handlers := append([]Subscriber{}, p.subs[channel]...)
	p.mu.RUnlock()
	for _, handler := range handlers {
		handler(payload)
	}
	return len(handlers)
}

File: python/cache/pubsub.py
from __future__ import annotations

from typing import Any, Callable, Dict, List


class PubSub:
    def __init__(self) -> None:
        self._subscribers: Dict[str, List[Callable[[Any], None]]] = {}

    def subscribe(self, channel: str, handler: Callable[[Any], None]) -> Callable[[], None]:
        handlers = self._subscribers.setdefault(channel, [])
        handlers.append(handler)

        def unsubscribe() -> None:
            handlers = self._subscribers.get(channel, [])
            if handler in handlers:
                handlers.remove(handler)

        return unsubscribe

    def publish(self, channel: str, payload: Any) -> int:
        handlers = list(self._subscribers.get(channel, []))
        for handler in handlers:
            handler(payload)
        return len(handlers)
